File: src/core/Options/Handlers/Chain.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using core.Account;
using core.Adapters.Options;
using core.Shared;
using core.Shared.Adapters.Brokerage;
using MediatR;

namespace core.Options
{
    public class Chain
    {
        public class Query : RequestWithUserId<CommandResponse<OptionDetailsViewModel>>
        {
            public Query(Ticker ticker, Guid userId) : base(userId)
            {
                Ticker = ticker;
            }

            public Ticker Ticker { get; }
        }

        public class Handler : IRequestHandler<Query, CommandResponse<OptionDetailsViewModel>>
        {
            private readonly IAccountStorage _accounts;
            private readonly IBrokerage _brokerage;
            
            public Handler(IAccountStorage accounts, IBrokerage brokerage)
            {
                _accounts = accounts;
                _brokerage = brokerage;
            }
            
            public async Task<CommandResponse<OptionDetailsViewModel>> Handle(Query request, CancellationToken cancellationToken)
            {
                var user = await _accounts.GetUser(request.UserId) ?? throw new InvalidOperationException("User not found");
                var priceResult = await _brokerage.GetQuote(user.State, request.Ticker);
                var price = priceResult.IsOk switch {
                    true => priceResult.Success.Price,
                    false => (decimal?)null
                };

                var detailsResponse = await _brokerage.GetOptions(user.State, request.Ticker);
                if (!detailsResponse.IsOk)
                {
                    return CommandResponse<OptionDetailsViewModel>.Failed(detailsResponse.Error.Message);
                }

                var model = MapOptionDetails(price, detailsResponse.Success!);

                return CommandResponse<OptionDetailsViewModel>.Success(model);
            }
        }

        public static OptionDetailsViewModel MapOptionDetails(
            decimal? price,
            OptionChain chain)
        {
            return new OptionDetailsViewModel
            {
                StockPrice = price,
                Options = chain.Options,
                Expirations = chain.Options.Select(o => o.ExpirationDate).Distinct().ToArray(),
                Volatility = chain.Volatility,
                NumberOfContracts = chain.NumberOfContracts,
            };
        }
    }
}

File: src/core.fs/Options/Chain.fs
module core.fs.Options.Chain

    open System
    open core.Account
    open core.Options
    open core.Shared
    open core.Shared.Adapters.Brokerage
    open core.fs
    
    type Query(ticker:string, userId:Guid) =
        member this.Ticker = ticker
        member this.UserId = userId
        
    type Handler(accounts:IAccountStorage, brokerage:IBrokerage) =
        interface IApplicationService
        
        member this.Handle (query:Query) = task {
            let! user = accounts.GetUser(userId=query.UserId)
            
            match user with
            | null -> return ServiceResponse<OptionDetailsViewModel>(ServiceError("User not found"))
            | _ ->
                
                let! priceResult = brokerage.GetQuote(state=user.State, ticker=query.Ticker)
                let price = match priceResult.IsOk with
                            | true -> Nullable<decimal>(priceResult.Success.Price)
                            | false -> Nullable<decimal>()
                            
                let! details = brokerage.GetOptions(state=user.State, ticker=query.Ticker)
                
                match details.IsOk with
                | true -> 
                    let model = OptionDetailsViewModel(price, details.Success)
                    return ServiceResponse<OptionDetailsViewModel>(model)
                | false ->
                    return ServiceResponse<OptionDetailsViewModel>(ServiceError(details.Error.Message))
        }
            